Source/WebCore/rendering/RenderLayer.cpp

```diff
@@ -3749,21 +3749,28 @@ void RenderLayer::paintLayerContents(GraphicsContext& context, const LayerPainti
         return isPaintingCompositedBackground;
     }();
 
-    if (shouldPaintContent && paintingInfo.subtreePaintRoot) {
-        RenderLayer* subtreeRootLayer = paintingInfo.subtreePaintRoot->enclosingLayer();
+    if (paintingInfo.paintBehavior.contains(PaintBehavior::DraggableSnapshot) && paintingInfo.subtreePaintRoot) {
+        if (paintingInfo.subtreePaintRoot->hasLayer()) {
+            CheckedPtr subtreeRootLayer = paintingInfo.subtreePaintRoot->enclosingLayer();
 
-        if (subtreeRootLayer) {
             bool isLayerInSubtree = (this == subtreeRootLayer) || isDescendantOf(*subtreeRootLayer);
 
             if (isLayerInSubtree) {
-                if (paintingInfo.subtreePaintRoot != &renderer()) {
+                if (&renderer() != paintingInfo.subtreePaintRoot) {
                     if (CheckedPtr rootAsBlock = dynamicDowncast<RenderBlock>(paintingInfo.subtreePaintRoot)) {
                         if (!rootAsBlock->isContainingBlockAncestorFor(renderer()))
                             shouldPaintContent = false;
                     }
                 }
             } else
                 shouldPaintContent = false;
+        } else {
+            if (renderer().isAbsolutelyPositioned() && &renderer() != paintingInfo.subtreePaintRoot) {
+                if (CheckedPtr rootAsBlock = dynamicDowncast<RenderBlock>(paintingInfo.subtreePaintRoot)) {
```

Comment:
I think this branch also behaves correctly for the case where `paintingInfo.subtreePaintRoot->hasLayer()` is true, so consider removing the first branch.

If you do keep both, factor the `isContainingBlockAncestorFor` chunk into a lambda to avoid repetition.